libraries/AP_Winch/AP_Winch_Daiwa.cpp

```diff
@@ -0,0 +1,227 @@
+    if (!SRV_Channels::function_assigned(SRV_Channel::k_winch)) {
+        return;
+    }
+
+    // read latest data from winch
+    read_data_from_winch();
+
+    // read pilot input
+    read_pilot_desired_rate();
+
+    // send outputs to winch
+    control_winch();
+}
+
+//send generator status
+void AP_Winch_Daiwa::send_status(const GCS_MAVLINK &channel)
+{
+    // prepare status bitmask
+    uint32_t status_bitmask = 0;
+    if (healthy()) {
+        status_bitmask |= MAV_WINCH_STATUS_HEALTHY;
+    }
+    if (latest.thread_end) {
+        status_bitmask |= MAV_WINCH_STATUS_FULLY_RETRACTED;
+    }
+    if (latest.moving > 0) {
+        status_bitmask |= MAV_WINCH_STATUS_MOVING;
+    }
+    if (latest.clutch > 0) {
+        status_bitmask |= MAV_WINCH_STATUS_CLUTCH_ENGAGED;
+    }
+
+    // convert speed percentage to absolute speed
+    const float speed_ms = fabsf(config.rate_max) * (float)latest.speed_pct;
+
+    // send status
+    mavlink_msg_winch_status_send(
+        channel.get_chan(),
+        AP_HAL::micros64(),
+        latest.line_length,
+        speed_ms,
+        (float)latest.tension_corrected * 0.01f,
+        latest.voltage,
+        latest.current,
+        latest.motor_temp,
+        status_bitmask);
+}
+
+// write log
+void AP_Winch_Daiwa::write_log()
+{
+    AP::logger().Write_Winch(
+            healthy(),
+            latest.thread_end,
+            latest.moving,
+            latest.clutch,
+            (uint8_t)config.control_mode,
+            config.length_desired,
+            get_current_length(),
+            config.rate_desired,
+            latest.tension_corrected,
+            latest.voltage,
+            constrain_int16(latest.motor_temp, INT8_MIN, INT8_MAX));
+}
+
+// read incoming data from winch and update intermediate and latest structures
+void AP_Winch_Daiwa::read_data_from_winch()
+{
+    // return immediately if serial port is not configured
+    if (uart == nullptr) {
+        return;
+    }
+
+    // read any available characters from serial port and send to GCS
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        int16_t b = uart->read();
+
+        if ((b >= '0' && b <= '9') || (b >= 'A' && b <= 'F') || (b >= 'a' && b <= 'f')) {
+            // add digits to buffer
+            buff[buff_len] = b;
+            buff_len++;
+            if (buff_len >= buff_len_max) {
+                buff_len = 0;
+                parse_state = ParseState::WAITING_FOR_TIME;
+            }
+        } else if (b == ',' || b == '\r') {
+            // comma or carriage return signals end of current value
+            buff[buff_len] = '\0';
+            long int ret = (int32_t)strtol(buff, nullptr, 16);
+            if (ret >= (long)INT32_MAX || ret <= (long)INT32_MIN) {
+                // failed to get valid number, throw away packet
+                parse_state = ParseState::WAITING_FOR_TIME;
+            } else {
+                // parse number received and empty buffer
```

Comment:
Part of comment incorrect / other part is kind of redundant :-)

Reply:
moved comment up a couple of lines so it makes more sense.